Jetpack Debug Helper: Data Mocker

sergeymitr opened the pull request:
#### Changes proposed in this Pull Request:

The new "Mocker" debugger module creates mock data for testing purposes.
There are two mockers implemented:
- Options to generate random options.
- Nonces, to create nonces with different timestamps.

The mocker works asynchronously, sending 5 requests at a time, each batch inserts up to 10k records into the database.

The PR also includes the new `loaderButton` JavaScript component extracted from the REST API Tester UI.
It's been reused for the Mocker UI.

#### Does this pull request change what data or activity we track or use?
No.

#### Testing instructions:
1. Disable WP Cron by defining the constant `define( 'DISABLE_WP_CRON', true );` so mocked data wouldn't be immediately removed.
2. Go to the Jetpack Debug (`/wp-admin/admin.php?page=jetpack-debug-tools`), activate Mocker.
3. Go to "Jetpack Debug -> Mocker", create 100000 mock "Options". Confirm that only 5 batches are running at a time, with 10 batches total.
4. Confirm that 100,000 mock options were inserted (`yarn docker:db` or `wp db cli`):
`select count(*) from wp_options where option_name like 'mock\_option\_%';`
5. Remove the options: `delete from wp_options where option_name like 'mock\_option\_%';`
6. Create 100000 mock "Nonces". Confirm that only 5 batches are running at a time, with 10 batches total.
7. Confirm that 100,000 mock nonces were inserted: `select count(*) from wp_options where option_name like 'jetpack\_nonce\_%';`
8. Remove the nonces: `delete from wp_options where option_name like 'jetpack\_nonce\_%';`
9. Don't forget to remove the `DISABLE_WP_CRON` constant 🙂 

#### Proposed changelog entry for your changes:
* Data mocking module for the Jetpack Debugger plugin.

sergeymitr committed:
Jetpack Debugger: Loader component.

- Created a separate "loader" component that can be reused between debugger modules.
- Modifying the API Tester to use that new loader component.

sergeymitr committed:
Jetpack Debugger: Mocker functionality.

The new "Mocker" debugger module creates mock data for testing purposes. There are two mockers implemented:
- Options to generate random options.
- Nonces, to create nonces with different timestamps.

sergeymitr committed:
Jetpack Debugger: Asynchronous Mocker.

The mocker JavaScript code has been rebuilt to work asynchronously.
It runs 5 batches at a time, each batch inserts up to 10k records.

bisko commented on packages/debug-helper/modules/inc/mockers/class-nonces-runner.php:
```diff
@@ -0,0 +1,74 @@
+			$sql                 .= "('" . esc_sql( $name ) . "', '" . esc_sql( $value ) . "', 'no'), ";
+		}
+
+		// phpcs:ignore WordPress.DB.PreparedSQL.NotPrepared
```
This can be moved to a prepared statement. It might look something like this: (Disclaimer: haven't tested it in a real env, but should give a good idea)

```php

$sql = "INSERT INTO {$wpdb->prefix}options (option_name, option_value, autoload) VALUES ";
$sql .= implode( ', ', array_fill( 0, $limit, '( %s, %s, "no" )' ) );

$values_to_insert = array();
for ( $i = 0; $i < $limit; ++$i ) {
    $values_to_insert = array_merge( $values_to_insert, $this->get_random_nonce() );
}

$prepared_query = $wpdb->prepare( $sql, $values_to_insert );

$wpdb->query( $prepared_query );

```

sergeymitr replied:
Thanks @bisko.
I was concerned about the performance, and after measuring the timing it turned out that `$wpdb->prepare()` is quite fast, so the difference is negligible.

Fixed now.

bisko replied:
Awesome! Can we remove the `phpcs:ignore` statement too? :)

sergeymitr replied:
Unfortunately, the `NotPrepared` error is still thrown, so we need it.
Apparently because the SQL code is passed via a variable.

We probably need to address that linting problem later on.

bisko commented on packages/debug-helper/modules/inc/mockers/class-nonces-runner.php:
```diff
@@ -0,0 +1,74 @@
+	 * @return array [ $option_key, $option_value ]
+	 */
+	private function get_random_nonce() {
+		// phpcs:ignore WordPress.WP.AlternativeFunctions.rand_rand
```
Why not `wp_rand()` ? Adding so many `phpcs` exceptions might create precedents that will be hard to clear out in the future.

sergeymitr replied:
`wp_rand()` is too slow, and we don't really care about the results being unpredictable.
In fact, switching to `wp_rand()` makes the whole process run 250% to 500% slower than if we were using `rand()`.
It's called multiple times for each record generated, so the difference quickly adds up if you need to insert hundreds of thousands of records.

bisko replied:
Do we care for the values being random? Why not use a simple decreasing counter from `time()` with some step?

Would that make it too predictable for our testing?

sergeymitr replied:
The string can be predictable, as long as the string format matches.
However, by passing the timestamp into the `get_random_string()` function we would add unnecessary complexity, while `rand()` does the job just fine.

bisko commented on packages/debug-helper/modules/inc/mockers/class-options-runner.php:
```diff
@@ -0,0 +1,73 @@
+			$sql .= "('" . esc_sql( $this->get_random_name() ) . "', '" . esc_sql( $this->get_random_value() ) . "', 'no'), ";
+		}
+
+		// phpcs:ignore WordPress.DB.PreparedSQL.NotPrepared
```
Same as my comment in the `run_batch` method.

bisko commented on packages/debug-helper/modules/inc/mockers/class-tools.php:
```diff
@@ -0,0 +1,39 @@
+	 * @return string
+	 */
+	public static function get_random_string( $length = 15, $length_max = null ) {
+		// phpcs:ignore WordPress.WP.AlternativeFunctions.rand_rand
```
Same as other comments about why not using `wp_rand()`

sergeymitr committed:
Jetpack Debugger: Mocker refactoring.

- Replaced direct SQL requests with prepared ones.
- Fixed an issue with the REST endpoint URL.

bisko reviewed (approved):
LGTM! Definitely the fastest mocker in town ;)
